File: server/app/simulation/race_engine/horse_types.py
"""Port of HorseTypes.ts."""

from __future__ import annotations

from enum import IntEnum


class Strategy(IntEnum):
    NIGE = 1
    SENKOU = 2
    SASI = 3
    OIKOMI = 4
    OONIGE = 5


class Aptitude(IntEnum):
    S = 0
    A = 1
    B = 2
    C = 3
    D = 4
    E = 5
    F = 6
    G = 7


class HorseParameters:
    """Mutable: RaceSolver clones this per-solver and green (stat-boost)
    skills mutate it in place mid-race, exactly like the TS `horse` field."""

    __slots__ = (
        "speed", "stamina", "power", "guts", "wisdom", "strategy",
        "distance_aptitude", "surface_aptitude", "strategy_aptitude", "raw_stamina",
    )

    def __init__(self, speed, stamina, power, guts, wisdom, strategy,
                 distance_aptitude, surface_aptitude, strategy_aptitude, raw_stamina):
        self.speed = speed
        self.stamina = stamina
        self.power = power
        self.guts = guts
        self.wisdom = wisdom
        self.strategy = strategy
        self.distance_aptitude = distance_aptitude
        self.surface_aptitude = surface_aptitude
        self.strategy_aptitude = strategy_aptitude
        self.raw_stamina = raw_stamina

    def clone(self) -> "HorseParameters":
        return HorseParameters(
            self.speed, self.stamina, self.power, self.guts, self.wisdom, self.strategy,
            self.distance_aptitude, self.surface_aptitude, self.strategy_aptitude, self.raw_stamina,
        )


def strategy_matches(s1: int, s2: int) -> bool:
    return s1 == s2 or (s1 == Strategy.NIGE and s2 == Strategy.OONIGE) or (s1 == Strategy.OONIGE and s2 == Strategy.NIGE)


def parse_strategy(s) -> Strategy:
    if not isinstance(s, str):
        return Strategy(s)
    u = s.upper()
    if u == "NIGE":
        return Strategy.NIGE
    if u == "SENKOU":
        return Strategy.SENKOU
    if u in ("SASI", "SASHI"):
        return Strategy.SASI
    if u == "OIKOMI":
        return Strategy.OIKOMI
    if u == "OONIGE":
        return Strategy.OONIGE
    raise ValueError("Invalid running strategy.")


def parse_aptitude(a, kind: str) -> Aptitude:
    if not isinstance(a, str):
        return Aptitude(a)
    u = a.upper()
    if u in Aptitude.__members__:
        return Aptitude[u]
    raise ValueError(f"Invalid {kind} aptitude.")

File: server/race_engine_rs/src/horse_types.rs
//! Port of horse_types.py.

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Strategy {
    Nige = 1,
    Senkou = 2,
    Sasi = 3,
    Oikomi = 4,
    Oonige = 5,
}

impl Strategy {
    pub fn from_i32(v: i32) -> Result<Strategy, String> {
        match v {
            1 => Ok(Strategy::Nige),
            2 => Ok(Strategy::Senkou),
            3 => Ok(Strategy::Sasi),
            4 => Ok(Strategy::Oikomi),
            5 => Ok(Strategy::Oonige),
            _ => Err(format!("Invalid running strategy int {v}.")),
        }
    }

    pub fn parse_str(s: &str) -> Result<Strategy, String> {
        match s.to_uppercase().as_str() {
            "NIGE" => Ok(Strategy::Nige),
            "SENKOU" => Ok(Strategy::Senkou),
            "SASI" | "SASHI" => Ok(Strategy::Sasi),
            "OIKOMI" => Ok(Strategy::Oikomi),
            "OONIGE" => Ok(Strategy::Oonige),
            _ => Err("Invalid running strategy.".to_string()),
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum Aptitude {
    S = 0,
    A = 1,
    B = 2,
    C = 3,
    D = 4,
    E = 5,
    F = 6,
    G = 7,
}

impl Aptitude {
    pub fn from_i32(v: i32) -> Result<Aptitude, String> {
        match v {
            0 => Ok(Aptitude::S),
            1 => Ok(Aptitude::A),
            2 => Ok(Aptitude::B),
            3 => Ok(Aptitude::C),
            4 => Ok(Aptitude::D),
            5 => Ok(Aptitude::E),
            6 => Ok(Aptitude::F),
            7 => Ok(Aptitude::G),
            _ => Err(format!("Invalid aptitude int {v}.")),
        }
    }

    pub fn parse_str(s: &str, kind: &str) -> Result<Aptitude, String> {
        match s.to_uppercase().as_str() {
            "S" => Ok(Aptitude::S),
            "A" => Ok(Aptitude::A),
            "B" => Ok(Aptitude::B),
            "C" => Ok(Aptitude::C),
            "D" => Ok(Aptitude::D),
            "E" => Ok(Aptitude::E),
            "F" => Ok(Aptitude::F),
            "G" => Ok(Aptitude::G),
            _ => Err(format!("Invalid {kind} aptitude.")),
        }
    }
}

/// Mutable: RaceSolver clones this per-solver and green (stat-boost) skills
/// mutate it in place mid-race, exactly like the Python/TS `horse` field.
#[derive(Clone, Debug)]
pub struct HorseParameters {
    pub speed: f64,
    pub stamina: f64,
    pub power: f64,
    pub guts: f64,
    pub wisdom: f64,
    pub strategy: Strategy,
    pub distance_aptitude: Aptitude,
    pub surface_aptitude: Aptitude,
    pub strategy_aptitude: Aptitude,
    pub raw_stamina: f64,
}

pub fn strategy_matches(s1: Strategy, s2: Strategy) -> bool {
    s1 == s2
        || (s1 == Strategy::Nige && s2 == Strategy::Oonige)
        || (s1 == Strategy::Oonige && s2 == Strategy::Nige)
}
